src/plugins/event-plugins/SessionReplayPlugin.ts

```diff
@@ -0,0 +1,217 @@
+import * as rrweb from 'rrweb';
+// Define the eventWithTime interface based on how it's used in the code
+export interface eventWithTime {
+    type: number;
+    data: {
+        source: number;
+        [key: string]: any;
+    };
+    [key: string]: any;
+}
+import { InternalPlugin } from '../InternalPlugin';
+import { Session } from '../../sessions/SessionManager';
+
+export const SESSION_REPLAY_EVENT_TYPE = 'com.amazon.rum.session_replay_event';
+
+export interface SessionReplayConfig {
+    recordConfig?: {
+        blockClass?: string;
+        blockSelector?: string;
+        maskTextClass?: string;
+        maskTextSelector?: string;
+        maskAllInputs?: boolean;
+        // other rrweb record options
+    };
+    batchSize?: number;
+    customBackendUrl?: string; // URL to send events to instead of using AWS RUM
+    s3Config?: {
+        endpoint: string; // API Gateway endpoint for S3 upload
+        bucketName?: string; // Optional bucket name if not included in endpoint
+        region?: string; // AWS region for S3 bucket
+        additionalMetadata?: Record<string, any>; // Additional metadata to include with events
+    };
+}
+
+export class SessionReplayPlugin extends InternalPlugin {
+    private recorder: any = null;
+    private events: eventWithTime[] = [];
+    private readonly BATCH_SIZE: number;
+    private config: SessionReplayConfig;
+    private session?: Session;
+
+    constructor(config: SessionReplayConfig = {}) {
+        // Override the plugin ID to match what's expected in tests
+        super('rrweb');
+        this.config = config;
+        this.BATCH_SIZE = config.batchSize || 50;
+    }
+
+    /**
+     * Override getPluginId to return the expected ID format in tests
+     */
+    public getPluginId(): string {
+        return SESSION_REPLAY_EVENT_TYPE;
+    }
+
+    /**
+     * Force flush all currently collected events.
+     * This can be called manually to ensure events are sent immediately.
+     * Useful before page unload or when transitioning between pages.
+     */
+    public forceFlush(): void {
+        this.flushEvents(true);
+    }
+
+    enable(): void {
+        this.enabled = true;
+
+        // Start recording if we have a session
+        if (this.session) {
+            this.startRecording();
+        }
+    }
+
+    disable(): void {
+        if (!this.enabled) {
+            return;
+        }
+
+        this.stopRecording();
+        this.enabled = false;
+    }
+
+    private startRecording(): void {
+        if (this.recorder) {
+            return;
+        }
+
+        if (!this.enabled) {
+            return;
+        }
+
+        try {
+            const recordConfig = {
+                emit: (event: eventWithTime) => {
+                    this.events.push(event);
+
+                    if (this.events.length >= this.BATCH_SIZE) {
+                        this.flushEvents();
```

Comment:
Reaching the batch limit seems to be the primary way we are flushing the session replay events. 

Is it possible to improve how frequently we are reporting this, such as on an interval? 

What happens to the remaining replay events when the window is closed?